File: Search/H-IndexII.java
/*
Follow up for H-Index: What if the citations array is sorted in ascending order? Could you optimize your algorithm?
*/
public class Solution {
    public int hIndex(int[] citations) {
        int lo = 0;
        int len = citations.length;
        int hi = len - 1;
        while(lo <= hi){
            int mid = (hi - lo) / 2 + lo;
            if(len - mid == citations[mid]) return citations[mid];
            if(len - mid < citations[mid]) hi = mid - 1;
            else lo = mid + 1;
        }
        return len - lo;
    }
}

File: Search/H-IndexII.js
/*
Follow up for H-Index: What if the citations array is sorted in ascending order? Could you optimize your algorithm?
*/

/**
 * @param {number[]} citations
 * @return {number}
 */
var hIndex = function(citations) {
    if(citations.length === 0) return 0;
    var len = citations.length;
    var lo = 0; var hi = len - 1;
    while(lo <= hi){
        var mid = Math.floor((hi - lo) / 2) + lo;
        if(citations[mid] == len - mid) return citations[mid];
        if(citations[mid] > len - mid) hi = mid - 1;
        else lo = mid + 1;
    }
    return len - lo;
};